src/ziggurat/middleware/default.clj

```diff
@@ -5,35 +5,80 @@
+                                                          (= :struct-value type)
+                                                          (protobuf-struct->persistent-map val)
+                                                          (= :list-value type)
+                                                          (reduce #(conj %1 (flatten-value %2))
+                                                                  []
+                                                                  val)
+                                                          :else
+                                                          val))
+                                                      value)))
+        protobuf-struct-entry-flattener (fn [res entry]
+                                          (let [key   (keyword (:key entry))
+                                                value (:value entry)]
+                                            (if key
+                                              (assoc res key (protobuf-value-flattener value))
+                                              res)))]
+    (reduce protobuf-struct-entry-flattener
+            {}
+            fields)))
+
 (defn deserialize-message
   "This function takes in the message(proto Byte Array) and the proto-class and deserializes the proto ByteArray into a
   Clojure PersistentHashMap.
   Temporary logic for migration of services to Ziggurat V3.0
     If the message is of type map, the function just returns the map as it is. In older versions of Ziggurat (< 3.0) we stored
     the messages in deserialized formats in RabbitMQ and those messages can be processed by this function. So we have this logic here."
-  [message proto-class topic-entity-name]
-  (if-not (map? message) ;; TODO: we should have proper dispatch logic per message type (not like this)
-    (try
-      (let [proto-klass  (protodef/mapdef proto-class)
-            loaded-proto (protodef/parse proto-klass message)
-            proto-keys   (-> proto-klass
-                             protodef/mapdef->schema
-                             :fields
-                             keys)]
-        (select-keys loaded-proto proto-keys))
-      (catch Throwable e
-        (let [service-name      (:app-name (ziggurat-config))
-              additional-tags   {:topic_name topic-entity-name}
-              default-namespace "message-parsing"
-              metric-namespaces [service-name topic-entity-name default-namespace]
-              multi-namespaces  [metric-namespaces [default-namespace]]]
-          (sentry/report-error sentry-reporter e (str "Couldn't parse the message with proto - " proto-class))
-          (metrics/multi-ns-increment-count multi-namespaces "failed" additional-tags)
-          nil)))
-    message))
+  ([message proto-class topic-entity-name]
+   (deserialize-message message proto-class topic-entity-name false))
+  ([message proto-class topic-entity-name flatten-protobuf-struct?]
```

Comment:
Request you to make the changes (with tests) in batch_proto_deserializer [1] too. This uses the method `deserialize_message` and can benefit from this change.

[1] https://github.com/gojek/ziggurat/blob/master/src/ziggurat/middleware/batch/batch_proto_deserializer.clj

Reply:
Made requested changes